In [5]:
import numpy as np
from sklearn.datasets import load_iris
from sklearn.linear_model import Perceptron

In [2]:
!pip3 install threadpoolctl


ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
mlxtend 0.21.0 requires matplotlib>=3.0.0, which is not installed.

[notice] A new release of pip available: 22.2.2 -> 22.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [1]:
pip install matplotlib==3.0.0

  Using cached matplotlib-3.0.0.tar.gz (36.3 MB)
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
Note: you may need to restart the kernel to use updated packages.


  error: subprocess-exited-with-error
  
  python setup.py egg_info did not run successfully.
  exit code: 1
  
  [47 lines of output]
  Edit setup.cfg to change the build options
  
  BUILDING MATPLOTLIB
              matplotlib: yes [3.0.0]
                  python: yes [3.8.10 (default, May 19 2021, 13:12:57) [MSC
                          v.1916 64 bit (AMD64)]]
                platform: yes [win32]
  
  REQUIRED DEPENDENCIES AND EXTENSIONS
                   numpy: yes [version 1.20.2]
        install_requires: yes [handled by setuptools]
                  libagg: yes [pkg-config information for 'libagg' could not
                          be found. Using local copy.]
                freetype: no  [The C/C++ header for freetype
                          (freetype2\ft2build.h) could not be found.  You may
                          need to install the development package.]
                     png: no  [The C/C++ header for png (png.h) could not be
                          found.  

In [1]:
import tensorflow as tf
from tensorflow import keras

from tensorflow.keras.layers import BatchNormalization
tf.__version__
keras.__version__

'2.4.0'

In [2]:
fashion_mnist = keras.datasets.fashion_mnist
(X_train_full, y_train_full), (X_test, y_test) = fashion_mnist.load_data()

In [3]:
X_train_full.shape


(60000, 28, 28)

In [10]:

X_train_full.dtype

dtype('uint8')

In [4]:
X_valid, X_train = X_train_full[:5000] / 255.0, X_train_full[5000:] / 255.0
y_valid, y_train = y_train_full[:5000], y_train_full[5000:]

In [37]:
features_names = ["T-shirt/top", "Trouser", "Pullover", "Dress", "Coat",
"Sandal", "Shirt", "Sneaker", "Bag", "Ankle boot"]

In [38]:
features_names[y_train[0]]

'Coat'

# SEQUENTIAL API METHOD OF TRAINING

In [14]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, activation="relu"))
model.add(keras.layers.Dense(100, activation="relu"))
model.add(keras.layers.Dense(10, activation="softmax"))

In [15]:
model.summary()

Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten (Flatten)            (None, 784)               0         
_________________________________________________________________
dense (Dense)                (None, 300)               235500    
_________________________________________________________________
dense_1 (Dense)              (None, 100)               30100     
_________________________________________________________________
dense_2 (Dense)              (None, 10)                1010      
Total params: 266,610
Trainable params: 266,610
Non-trainable params: 0
_________________________________________________________________


In [31]:
model.compile(loss="sparse_categorical_crossentropy", optimizer="sgd", metrics=["accuracy"])

In [17]:
history = model.fit(X_train, y_train, epochs=30, validation_data=(X_valid, y_valid))

Epoch 1/30
1719/1719 [==============================] - 5s 3ms/step - loss: 0.7183 - accuracy: 0.7637 - val_loss: 0.5056 - val_accuracy: 0.8316
Epoch 2/30
1719/1719 [==============================] - 5s 3ms/step - loss: 0.4899 - accuracy: 0.8294 - val_loss: 0.4507 - val_accuracy: 0.8480
Epoch 3/30
1719/1719 [==============================] - 5s 3ms/step - loss: 0.4432 - accuracy: 0.8460 - val_loss: 0.4210 - val_accuracy: 0.8548
Epoch 4/30
1719/1719 [==============================] - 4s 3ms/step - loss: 0.4161 - accuracy: 0.8529 - val_loss: 0.4205 - val_accuracy: 0.8540
Epoch 5/30
1719/1719 [==============================] - 5s 3ms/step - loss: 0.3947 - accuracy: 0.8616 - val_loss: 0.4410 - val_accuracy: 0.8418
Epoch 6/30
1719/1719 [==============================] - 5s 3ms/step - loss: 0.3788 - accuracy: 0.8665 - val_loss: 0.3669 - val_accuracy: 0.8746
Epoch 7/30
1719/1719 [==============================] - 5s 3ms/step - loss: 0.3655 - accuracy: 0.8703 - val_loss: 0.3670 - val_accuracy:

# Here We have put batch size to 8

In [32]:
history = model.fit(X_train, y_train, epochs=20, batch_size=8)

Epoch 1/20
6875/6875 [==============================] - 12s 2ms/step - loss: 0.5518 - accuracy: 0.8083
Epoch 2/20
6875/6875 [==============================] - 12s 2ms/step - loss: 0.4061 - accuracy: 0.8534
Epoch 3/20
6875/6875 [==============================] - 13s 2ms/step - loss: 0.3622 - accuracy: 0.8673
Epoch 4/20
6875/6875 [==============================] - 12s 2ms/step - loss: 0.3351 - accuracy: 0.8770
Epoch 5/20
6875/6875 [==============================] - 16s 2ms/step - loss: 0.3166 - accuracy: 0.8841
Epoch 6/20
6875/6875 [==============================] - 18s 3ms/step - loss: 0.2997 - accuracy: 0.8893 0s - loss: 0
Epoch 7/20
6875/6875 [==============================] - 16s 2ms/step - loss: 0.2867 - accuracy: 0.8949
Epoch 8/20
6875/6875 [==============================] - 14s 2ms/step - loss: 0.2748 - accuracy: 0.9001
Epoch 9/20
6875/6875 [==============================] - 16s 2ms/step - loss: 0.2644 - accuracy: 0.9014
Epoch 10/20
6875/6875 [==============================] - 14s

In [33]:
model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 86.6045 - accuracy: 0.8272


[86.6044921875, 0.8271999955177307]

# Batch Normalization in sequential api

In [21]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, activation="relu"))
model.add(BatchNormalization())
model.add(keras.layers.Dense(100, activation="relu"))
model.add(BatchNormalization())
model.add(keras.layers.Dense(10, activation="softmax"))


In [22]:
model.summary()

Model: "sequential_3"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_3 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_9 (Dense)              (None, 300)               235500    
_________________________________________________________________
batch_normalization_5 (Batch (None, 300)               1200      
_________________________________________________________________
dense_10 (Dense)             (None, 100)               30100     
_________________________________________________________________
batch_normalization_6 (Batch (None, 100)               400       
_________________________________________________________________
dense_11 (Dense)             (None, 10)                1010      
Total params: 268,210
Trainable params: 267,410
Non-trainable params: 800
______________________________________________

In [23]:
model.compile(loss="sparse_categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

In [24]:
history = model.fit(X_train, y_train, epochs=10, batch_size=8)

Epoch 1/10
6875/6875 [==============================] - 18s 3ms/step - loss: 0.6188 - accuracy: 0.7845
Epoch 2/10
6875/6875 [==============================] - 18s 3ms/step - loss: 0.5041 - accuracy: 0.8201
Epoch 3/10
6875/6875 [==============================] - 21s 3ms/step - loss: 0.4663 - accuracy: 0.8354
Epoch 4/10
6875/6875 [==============================] - 22s 3ms/step - loss: 0.4369 - accuracy: 0.8439
Epoch 5/10
6875/6875 [==============================] - 18s 3ms/step - loss: 0.4156 - accuracy: 0.8490
Epoch 6/10
6875/6875 [==============================] - ETA: 0s - loss: 0.4119 - accuracy: 0.85 - 19s 3ms/step - loss: 0.4119 - accuracy: 0.8507
Epoch 7/10
6875/6875 [==============================] - 19s 3ms/step - loss: 0.3990 - accuracy: 0.8561
Epoch 8/10
6875/6875 [==============================] - 18s 3ms/step - loss: 0.3955 - accuracy: 0.8574 0s -
Epoch 9/10
6875/6875 [==============================] - 18s 3ms/step - loss: 0.3810 - accuracy: 0.8609
Epoch 10/10
6875/6875 [===

In [36]:
model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 2ms/step - loss: 66.9301 - accuracy: 0.8591


[66.9300765991211, 0.8590999841690063]

# Adding Dropout layers in Sequential API

In [28]:
model = keras.models.Sequential()
model.add(keras.layers.Flatten(input_shape=[28, 28]))
model.add(keras.layers.Dense(300, activation="relu"))
keras.layers.Dropout(0.5)
model.add(keras.layers.Dense(100, activation="relu"))
keras.layers.Dropout(0.5)
model.add(keras.layers.Dense(10, activation="softmax"))

In [32]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
flatten_5 (Flatten)          (None, 784)               0         
_________________________________________________________________
dense_15 (Dense)             (None, 300)               235500    
_________________________________________________________________
dense_16 (Dense)             (None, 100)               30100     
_________________________________________________________________
dense_17 (Dense)             (None, 10)                1010      
Total params: 266,610
Trainable params: 266,610
Non-trainable params: 0
_________________________________________________________________


In [33]:
model.compile(loss="sparse_categorical_crossentropy", optimizer="adam", metrics=["accuracy"])

In [35]:
history = model.fit(X_train, y_train, epochs=10, batch_size=8)

Epoch 1/10
6875/6875 [==============================] - 16s 2ms/step - loss: 0.4865 - accuracy: 0.8222
Epoch 2/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.3713 - accuracy: 0.8625
Epoch 3/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.3364 - accuracy: 0.8741
Epoch 4/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.3139 - accuracy: 0.8831
Epoch 5/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.2974 - accuracy: 0.8895
Epoch 6/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.2852 - accuracy: 0.8935
Epoch 7/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.2748 - accuracy: 0.8967
Epoch 8/10
6875/6875 [==============================] - 15s 2ms/step - loss: 0.2692 - accuracy: 0.9006
Epoch 9/10
6875/6875 [==============================] - 16s 2ms/step - loss: 0.2574 - accuracy: 0.9042
Epoch 10/10
6875/6875 [==============================] - 16s 2ms/step - l

In [18]:
model.evaluate(X_test, y_test)

313/313 [==============================] - 1s 3ms/step - loss: 91.6958 - accuracy: 0.8145


[91.69583129882812, 0.8144999742507935]

# functional api method

In [47]:
input_ = keras.layers.Input(shape=[28, 28])
flatten = keras.layers.Flatten(input_shape=[28, 28])(input_)
hidden1 = keras.layers.Dense(2**14, activation="relu")(flatten)
hidden2 = keras.layers.Dense(512, activation='relu')(hidden1)
hidden3 = keras.layers.Dense(28*28, activation='relu')(hidden2)
reshap = keras.layers.Reshape((28, 28))(hidden3)
concat_ = keras.layers.Concatenate()([input_, reshap])
flatten2 = keras.layers.Flatten(input_shape=[28, 28])(concat_)
output = keras.layers.Dense(10, activation='softmax')(flatten2)
model = keras.Model(inputs=[input_], outputs=[output] )

In [48]:
model.summary()

Model: "functional_5"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_4 (InputLayer)            [(None, 28, 28)]     0                                            
__________________________________________________________________________________________________
flatten_2 (Flatten)             (None, 784)          0           input_4[0][0]                    
__________________________________________________________________________________________________
dense_14 (Dense)                (None, 16384)        12861440    flatten_2[0][0]                  
__________________________________________________________________________________________________
dense_15 (Dense)                (None, 512)          8389120     dense_14[0][0]                   
_______________________________________________________________________________________

In [51]:
model.compile(loss="sparse_categorical_crossentropy", optimizer="adam",
              metrics=["accuracy"])
h = model.fit(X_train, y_train, epochs=8, batch_size=8)

Epoch 1/8
6875/6875 [==============================] - 819s 119ms/step - loss: 0.4732 - accuracy: 0.8354
Epoch 2/8
6875/6875 [==============================] - 800s 116ms/step - loss: 0.4386 - accuracy: 0.8459
Epoch 3/8
6875/6875 [==============================] - 874s 127ms/step - loss: 0.4263 - accuracy: 0.8512
Epoch 4/8
6875/6875 [==============================] - 810s 118ms/step - loss: 0.4169 - accuracy: 0.8557
Epoch 5/8
6875/6875 [==============================] - 776s 113ms/step - loss: 0.4120 - accuracy: 0.8559
Epoch 6/8
6875/6875 [==============================] - 775s 113ms/step - loss: 0.4071 - accuracy: 0.8590
Epoch 7/8
6875/6875 [==============================] - 775s 113ms/step - loss: 0.4049 - accuracy: 0.8590
Epoch 8/8
6875/6875 [==============================] - 763s 111ms/step - loss: 0.4005 - accuracy: 0.8608


# Here we have done batch normalization to functional api

In [7]:
X_train_full=X_train_full.reshape(60000,784)/255.0

testdata=X_test.reshape(10000,784)/255.0

In [15]:
Layer1=keras.layers.Input(shape=(784,))
Layer2=keras.layers.Dense(80,activation="relu")(Layer1)
Layer3=keras.layers.Dense(320,activation="relu")(Layer2)
concat=keras.layers.Concatenate()([Layer1,Layer2])

BN=keras.layers.BatchNormalization()(concat)
op=keras.layers.Dense(10,activation="softmax")(BN)

model=keras.Model(inputs=[Layer1],outputs=[op])

In [14]:
model.summary()

Model: "functional_1"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_1 (InputLayer)            [(None, 784)]        0                                            
__________________________________________________________________________________________________
dense (Dense)                   (None, 80)           62800       input_1[0][0]                    
__________________________________________________________________________________________________
concatenate (Concatenate)       (None, 864)          0           input_1[0][0]                    
                                                                 dense[0][0]                      
__________________________________________________________________________________________________
batch_normalization (BatchNorma (None, 864)          3456        concatenate[0][0]     

In [18]:
model.compile(loss="sparse_categorical_crossentropy",optimizer="adam",metrics=["accuracy"])

opdata=model.fit(X_train_full,y_train_full,epochs=20,validation_split=0.40)

Epoch 1/20
1125/1125 [==============================] - 5s 5ms/step - loss: 0.2933 - accuracy: 0.8903 - val_loss: 0.3461 - val_accuracy: 0.8782
Epoch 2/20
1125/1125 [==============================] - 4s 4ms/step - loss: 0.2899 - accuracy: 0.8915 - val_loss: 0.3439 - val_accuracy: 0.8802
Epoch 3/20
1125/1125 [==============================] - 4s 4ms/step - loss: 0.2897 - accuracy: 0.8925 - val_loss: 0.3463 - val_accuracy: 0.8787
Epoch 4/20
1125/1125 [==============================] - 4s 4ms/step - loss: 0.2851 - accuracy: 0.8933 - val_loss: 0.3804 - val_accuracy: 0.8641
Epoch 5/20
1125/1125 [==============================] - 4s 4ms/step - loss: 0.2849 - accuracy: 0.8926 - val_loss: 0.3642 - val_accuracy: 0.8745
Epoch 6/20
1125/1125 [==============================] - 4s 4ms/step - loss: 0.2816 - accuracy: 0.8946 - val_loss: 0.3445 - val_accuracy: 0.8784
Epoch 7/20
1125/1125 [==============================] - 4s 4ms/step - loss: 0.2813 - accuracy: 0.8944 - val_loss: 0.3707 - val_accuracy:

# Here we performed both batch normalization and dropout on the model

In [19]:
Layer1=keras.layers.Input(shape=(784,))
Layer2=keras.layers.Dense(80,activation="relu")(Layer1)
Layer3=keras.layers.Dense(320,activation="relu")(Layer2)
concat=keras.layers.Concatenate()([Layer1,Layer2])

BN=keras.layers.BatchNormalization()(concat)

Drop=keras.layers.Dropout(0.5)(BN)
op=keras.layers.Dense(10,activation="softmax")(Drop)
model=keras.Model(inputs=[Layer1],outputs=[op])

In [20]:
model.summary()

Model: "functional_5"
__________________________________________________________________________________________________
Layer (type)                    Output Shape         Param #     Connected to                     
input_3 (InputLayer)            [(None, 784)]        0                                            
__________________________________________________________________________________________________
dense_6 (Dense)                 (None, 80)           62800       input_3[0][0]                    
__________________________________________________________________________________________________
concatenate_2 (Concatenate)     (None, 864)          0           input_3[0][0]                    
                                                                 dense_6[0][0]                    
__________________________________________________________________________________________________
batch_normalization_2 (BatchNor (None, 864)          3456        concatenate_2[0][0]   

In [21]:
model.compile(loss="sparse_categorical_crossentropy",optimizer="adam",metrics=["accuracy"])

opdata=model.fit(X_train_full,y_train_full,epochs=20,validation_split=0.30)

Epoch 1/20
1313/1313 [==============================] - 6s 5ms/step - loss: 0.7367 - accuracy: 0.7554 - val_loss: 0.5018 - val_accuracy: 0.8216
Epoch 2/20
1313/1313 [==============================] - 6s 4ms/step - loss: 0.5250 - accuracy: 0.8170 - val_loss: 0.4677 - val_accuracy: 0.8320
Epoch 3/20
1313/1313 [==============================] - 6s 4ms/step - loss: 0.4921 - accuracy: 0.8275 - val_loss: 0.4347 - val_accuracy: 0.8477
Epoch 4/20
1313/1313 [==============================] - 6s 4ms/step - loss: 0.4746 - accuracy: 0.8340 - val_loss: 0.4357 - val_accuracy: 0.8441
Epoch 5/20
1313/1313 [==============================] - 6s 4ms/step - loss: 0.4658 - accuracy: 0.8365 - val_loss: 0.4346 - val_accuracy: 0.8451
Epoch 6/20
1313/1313 [==============================] - 6s 5ms/step - loss: 0.4554 - accuracy: 0.8395 - val_loss: 0.4225 - val_accuracy: 0.8511
Epoch 7/20
1313/1313 [==============================] - 6s 4ms/step - loss: 0.4534 - accuracy: 0.8386 - val_loss: 0.4160 - val_accuracy:

PROCEDURE:

1.Here we are taking the MNIST fashion data set and using tensorflow libraries.

Tensorflow used for:
It is more powerful than the sequential API in the sense branching or sharing of layers is allowed here. And also it can have multiple inputs and outputs

Dataset:
Fashion-MNIST is a dataset of Zalando's article images—consisting of a training set of 60,000 examples and a test set of 10,000 examples. Each example is a 28x28 grayscale image, associated with a label from 10 classes. Zalando intends Fashion-MNIST to serve as a direct drop-in replacement for the original MNIST dataset for benchmarking machine learning algorithms. It shares the same image size and structure of training and testing splits.

2.There are two methods in training the DEEP LEARNING MODEL ,those are sequential method  and functional api

3.In sequential , we cannot have multiple input at a time , it is designed layer by layer

4.First we did batchnormalization and dropout layering for sequential methods, we got accuracy of 86 and 90 % respectively,

5.Functional api : It is more advanced than the sequential API in the sense branching or sharing of layers is allowed here and also it can have multiple inputs and outputs.

Optimizer:Here we used Adam optimizer, it is a replacement optimization algorithm for stochastic gradient descent for training deep learning models. 

6.After adding the batchnormaliation and drop out layers , we compile the model and fit the model to run epochs with set limit.


AActivation function: it defines the what node/nueron has to do .


7.So after the fit we got 90.8 % and 87% accuracy in functional api.



